src/service/auth/method/repository.rs

```diff
@@ -0,0 +1,979 @@
+use std::str::FromStr;
+
+use lib::database::LocalDbClientError;
+use lib::id_gen::{ConcurrentSnowflake, ConcurrentSnowflakeError};
+use uuid::Uuid;
+
+use gen::database::{
+    FunAuthAuthenticateReq, FunAuthAuthenticateResp, FunAuthAuthenticateRespRow,
+    FunAuthAuthorizeReq, FunAuthAuthorizeResp, FunAuthAuthorizeRespRow,
+    FunAuthBasicAuthenticateReq, FunAuthBasicAuthenticateResp, FunAuthBasicAuthenticateRespRow,
+    FunAuthChangePasswordReq, FunAuthChangePasswordResp, FunAuthChangePasswordRespRow,
+    FunAuthGetPasswordSaltReq, FunAuthGetPasswordSaltResp, FunAuthGetPasswordSaltRespRow,
+    FunAuthGetRecoveryQuestionsReq, FunAuthGetRecoveryQuestionsResp,
+    FunAuthGetRecoveryQuestionsRespRow, FunAuthResetPasswordReq, FunAuthResetPasswordResp,
+    FunAuthResetPasswordRespRow, FunAuthSetRecoveryQuestionsReq, FunAuthSetRecoveryQuestionsResp,
+    FunAuthSetRecoveryQuestionsRespRow, FunAuthSetTokenReq, FunAuthSetTokenResp,
+    FunAuthSetTokenRespRow, FunAuthSignupReq, FunAuthSignupResp, FunAuthSignupRespRow,
+    FunGetRecoveryQuestionDataReq, FunGetRecoveryQuestionDataResp,
+    FunGetRecoveryQuestionDataRespRow, FunSubmitRecoveryAnswersReq, FunSubmitRecoveryAnswersResp,
+    FunSubmitRecoveryAnswersRespRow,
+};
+use lib::database::LocalDbClient;
+
+use localdb::parsetools::*;
+
+use gen::model::{EnumRecoveryQuestionCategory, EnumRole, EnumService};
+
+pub async fn fun_auth_signup(
```

Comment:
DALs mixed with rust and SQL? Works with this demo, but not suitable in places where transactions are needed.
I understand that procedural functions are rarely implemented in small databases. Some devs write DAL with ORM, while we use plpgsql with code-generated bindings.

Reply:
I'm not sure I understand the alternative, I implemented the repository functions to mimic the previous procedural function behaviors as closely as possible using GlueSQL behind a websocket server, this meant using Rust to throw some Exceptions and to do assertions that were previously done in pure SQL

Also noteworthy are some hacky stuff like multi-request transactions (GlueSQL does not yet rollback a transaction on COMMIT if one of the statements threw an Exception, so I had to check transaction status and manually trigger the rollback when applicable), I'm not proud of it, but the challenge was to change the database into something that wouldn't need a container, and preferably rust based, and I did my best to do that

I totally get if you guys don't think this a good solution, though

That being said, if we decide that I am to continue working on this change the way it is, could you elaborate a little more on how you imagine this being done?

Reply:
The alternative means that
1. you put db function inside GlueSQL server
2. you call these db functions remotely
The current solution is that:
1. you implement a GlueSQL server
2. you run db functions in app server, which issues SQL to GlueSQL server

Reply:
In one day we'll finally roll our own database for all this kind of stuff, JK

Reply:
> In one day we'll finally roll our own database for all this kind of stuff, JK
Lol, you guys are insane, I love it

> The alternative means that
> 1. you put db function inside GlueSQL server
> 2. you call these db functions remotely(these handles/stubs are code generated)
> The current solution is that:
> 3. you implement a GlueSQL server
> 4. you run db functions in app server, which issues SQL to GlueSQL server
I think I get it, and It's pretty cool, but if I put these functions inside the server, even if I can expose some way to call them remotely, they will still be a mix of SQL and Rust since GlueSQL's SQL does not have features like if/else clauses, functions, throw Exception, or declare variable!

Or are you talking about implementing procedural functions in GlueSQL? I'm getting a feeling this change will get out of control quickly, lol

Reply:
> I think I get it, and It's pretty cool, but if I put these functions inside the server, even if I can expose some way to call them remotely, they will still be a mix of SQL and Rust since GlueSQL's SQL does not have features like if/else clauses, functions, throw Exception, or declare variable!
The mixture of SQL and Rust is unavoidable unless we implement GlueSQL somehow. Since this is a demo, we can have them mixed for now.